New resource - `azurerm_batch_job`

magodo opened the pull request:
Fixes #7862.

The test passes in TC:

![image](https://user-images.githubusercontent.com/7970698/125712158-50e10549-ae76-4837-8664-25990f9bd81f.png)

jackofallops reviewed (changes requested):
Thanks @magodo - A few comments / questions below, but I think looks mostly good. Does this resource have value / use without being able to configure / add tasks?

magodo commented on azurerm/internal/services/batch/batch_job.go:
```diff
@@ -0,0 +1,331 @@
+
+func (r BatchJobResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
```
Since this resource is quite fast to create/read/delete (almost instantly).

magodo commented on azurerm/internal/services/batch/batch_job.go:
```diff
@@ -0,0 +1,331 @@
+	BatchPoolId string `tfschema:"batch_pool_id"`
+	DisplayName string `tfschema:"display_name"`
+	Priority    int    `tfschema:"priority"`
+	// MaxWallClockTime  string `tfschema:"max_wall_clock_time"`
```
Agreed.

magodo commented on azurerm/internal/services/batch/batch_job.go:
```diff
@@ -0,0 +1,331 @@
+
+func (r BatchJobResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
```
Same reason as above.

magodo commented:
@jackofallops Thank you for the review! I've updated the comments above, please take another review.

A batch job is the prerequisite of a batch task, where the latter one resides inside the former. Since the ask from issue #7862 is all about a job, hence I submit this PR to resolve that. Maybe we can look into implementing the batch task in the future, while I feel due to the nature that a task is kind of short lived object in Azure, it might not be a good fit for Terraform.

katbyte commented on azurerm/internal/services/batch/batch_job.go:
```diff
@@ -0,0 +1,318 @@
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+		"common_environment_setting": {
```
```suggestion
		"common_environment_properties": {
```

magodo replied:
The `azurerm_batch_pool` used the term for [`environment`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/batch_pool#environment), shall we use that name to be consistent?

katbyte replied:
i would rather use the name with the better UX and leave a comment to change the old one in v3.0

katbyte commented on azurerm/internal/services/batch/batch_job.go:
```diff
@@ -0,0 +1,318 @@
+			Default:      0,
+			ValidateFunc: validation.IntBetween(-1000, 1000),
+		},
+		"max_task_retry_count": {
```
```suggestion
		"task_retry_maximum": {
```

magodo replied:
The `azurerm_batch_pool` used the same term for [`max_task_retry_count`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/batch_pool#max_task_retry_count), shall we keep this name to be consistent?

katbyte replied:
i would rather use the name with the better UX and leave a comment to change the old one in v3.0

magodo commented:
@katbyte Thank you for the comments! I've renamed both properties and also added TODO comment for the azurerm_batch_pool resource to remind renaming those properties in v3.0.

katbyte reviewed (approved):
Thanks @magodo - LGTM 🍔
